blog/_posts/2021-08-20-more-prefab-tools.md

```diff
@@ -0,0 +1,177 @@
+With this framework, it becomes simple to turn a C++ command line tool into web application.
+Running scientific software on the web has enormous potential to expand reach and engage the public since literally anyone can replicate your results and play with experimental conditions.
+However, just because something can run on the web does not mean it will work well.
+Developing a graphical user interface is a lot more work than making a command line one.
+
+To expedite porting from command line to browser, Empirical adds in simple components like divs, spans, and buttons and more complicated controls like a whole configuration panel for managing settings.
+Since these components are created and managed in C++ they can provide an graphical user interface to manage the execution of preexisting code.
+We call these components "prefabricated web tools" and they have be developed using the open source front-end toolkit, [Bootstrap](https://getbootstrap.com/docs/5.1/getting-started/introduction/).
+My predecessor Sarah Boyd developed a large number of components and you can read about her work [here]({{ site.baseurl }}/blog/prefab-tools.html).
+I took up the torch this summer to add to this previous work!
+
+## Goals
+The goal of the project this summer was to enhance existing prefabricated components and to create new ones.
+In addition, some prefabricated components had issues which needed to be addressed (:bug:).
+The following [issue](https://github.com/devosoft/Empirical/issues/334) listed some components to develop and along the way we thought up some more (:bulb:)! This [issue](https://github.com/devosoft/Empirical/issues/444) contains the most current list of project ideas for prefab components.
+
+#### "It will be simple," they said, "a button", they said
+Thinking it would be a relatively simple thing to start with, I began work adding a reload button to the `ConfigPanel` as one of my first tasks.
+Reloading the page allows us to add in a query string to the URL which contains key-value pairs that have the same effect as command line flags.
+Empirical's `ArgManager` class then handles interpreting both the command line flags and the web key-value pairs and adding them to a `Config` instance.
+
+However, this task almost immediately stalled on discovery of a bug!
+
+:bug: __Bug__: In an attempt to make Config objects flexible, `std::stringstream` was being used to extract `double`s, `float`s and `int`s from text entered by the user.
+This worked fine until you tried it with `std::string`s.
+The `std::stringstream` method resulted in only the first word being copied!
+
+:butterfly: __Solution__: Keeping a templated interface was necessary for efficiency and clarity. However, we needed different behavior for a specific type. Luckily since C++17 we have `constexpr` and type traits. Using
+```c++
+  if constexpr(std::is_same<MY_TYPE, std::>::value) {
+    // do something ...
+  }
+```
```

Comment:
I really like that you're using the markdown C++, I was initially using images of code but Jamie pointed out that they would have inconsistent sizes but putting the code in markdown would keep it consistent. Looks good!